Support tensor of indices as loop iter-arg in sequences of pointer arithmetic

nhat-nguyen opened the pull request:
This PR adds support for tensor of indices that are updated in each loop iteration while also being used in pointer arithmetic sequences.

## Approach

Similarly to the pointer types, in the PtrAnalysis pre-pass, we prematurely generate the `tts.get_structured_state` ops for tensor of integers. The important note here is we do not need to know whether these ops will eventually be used in a pointer arithmetic sequence. Any values that are not used in a pointer arithmetic sequence will be removed later in the process.

This approach can easily be extended to other kinds of values that might be used in pointer arithmetic sequences. At a high level, `tts.get_structured_state` can always be used to "wrap" a triton value. This op returns two kinds of values: the first value is always of the same type as the wrapped value, while the remaining values expose the important fields in `PtrState` that are necessary for codegen in scf.for.

The first return value of `tts.get_structured_state` is always an SSA value of the same type as the original value; users of the original triton value will then use this first return value from `tts.get_structured_state` instead. With this approach, even if the original triton value ends up not being used in a pointer arithmetic sequence, it is very easy to revert the IR to the original form by simply deleting the `tts.get_structured_state` op and forwarding the original triton value to its users again.

The other return values then expose the important fields in PtrState that are necessary to generate the code in loops (offsets and strides). Within a loop, for every wrapped triton value returned by a `tts.get_structured_state` op at index `i`, we can always get the corresponding offsets in each loop iteration at index `i + 1` and strides at index `i + 2`.

## Changes

+ Updated the pre-pass to insert `tts.get_structured_state` ops that wrap tensor of indices
+ With the introduction of tensor of indices in loops, we now have to manually visit the `tts.get_structured_state` ops to generate the ops for updating PtrState. We previously did not have to do this because triton pointers always have a `tt.addptr` at the end of each loop, right before yielding the values, which always triggers the process for generating the state-update ops
+ Logic for determining whether a loop iter-arg should have its PtrState updated is improved. We do a BFS-like scan starting from the return values of `tts.get_structured_state` ops to determine if an iter-arg originates from a value that may need its PtrState populated
+ Preliminary support for mask sequences being updated in a loop; this is a bit of a hack and will need more robust implementation if these use cases appear more frequently.
+ Add tests for various scenarios

nhat-nguyen commented:
@haishanzzzz Here's the PR for supporting tensor of indices in loop iter-args that are used in pointer sequence. Would appreciate if you could have a look! 😄

robbie-vanderzee-microsoft commented on lib/Dialect/TritonStructured/IR/TritonStructuredOps.cpp:
```diff
@@ -137,21 +136,24 @@ GetStructuredStateOp::getOffsetAndStrideTypes(MLIRContext *context,
-    if (auto ptrType =
-            dyn_cast<triton::PointerType>(tensorType.getElementType())) {
+  if (auto tensorType = llvm::dyn_cast<RankedTensorType>(type)) {
+    if (tensorType.getElementType().isIntOrIndex()) {
```
Ditto here, what causes us to only care about `IntOrIndex` types?

nhat-nguyen replied:
similarly to the other comment, only tensor of int or index can potentially be used to compute pointer offsets and hence we would only need to use the `tts.get_structured_state` ops for values of such types.

haishanzzzz commented on lib/AnalysisStructured/PtrAnalysis.cpp:
```diff
@@ -1131,6 +1146,86 @@ LogicalResult PtrAnalysis::rewriteLoadOp(triton::LoadOp op) {
+  std::queue<Value> q;
+  DenseSet<Value> visited;
+
+  op->walk([&q, &visited](tts::GetStructuredStateOp op) {
```
nit, we are overwriting the name `op`

haishanzzzz reviewed (approved):
PR LGTM. Thank you Nhat!

nhat-nguyen commented on lib/Conversion/TritonToStructured/TritonToStructuredPass.cpp:
```diff
@@ -286,9 +302,11 @@ class TritonToStructuredPass
-    if (failed(runTritonToStructuredPrepass())) {
-      signalPassFailure();
-      return;
+    if (!skipPrepass) {
```
yup will update in a separate PR with some more comments (i just want to get this in first 😄)

nhat-nguyen commented on lib/Conversion/TritonToStructured/TritonToStructuredPass.cpp:
```diff
@@ -84,7 +85,8 @@ class TritonToStructuredPass
                                 -> std::optional<LogicalResult> {
-      if (!isa<triton::PointerType>(tensorType.getElementType())) {
+      if (!isa<triton::PointerType>(tensorType.getElementType()) &&
+          !tensorType.getElementType().isIntOrIndex()) {
```
this is because only tensor of int or index can potentially be used to compute pointer offsets; i will add some more comments in a separate PR (just want to get this in first)

nhat-nguyen commented:
@robbie-vanderzee-microsoft I just pushed another commit that addresses all your comments. Would you mind resetting your review and giving another approval? Thanks!

robbie-vanderzee-microsoft reviewed (commented):
Looks good. Thanks for clarifying!
